axiom/tests/test_php_packages.c

```diff
@@ -0,0 +1,189 @@
+  package3 = nr_php_package_create("Package Three", "12.3.0");
+  /* Should not crash: */
+  nr_php_packages_add_package(NULL, package1);
+  nr_php_packages_add_package(hm, NULL);
+  nr_php_packages_add_package(hm, package1);
+  nr_php_packages_add_package(hm, package2);
+  nr_php_packages_add_package(hm, package3);
+
+  count = nr_php_packages_count(hm);
+
+  tlib_pass_if_int_equal("package count", 3, count);
+
+  nr_php_packages_destroy(&hm);
+  tlib_pass_if_null("PHP packages hashmap destroyed", hm);
+}
+
+static void test_php_package_to_json(void) {
+  char* json;
+  nr_php_package_t* package1;
+
+  // Test: convert package to json
+  package1 = nr_php_package_create("TestPackage", "7.2.0");
+  json = nr_php_package_to_json(package1);
+  tlib_pass_if_str_equal("valid package", "[\"TestPackage\",\"7.2.0\",{}]",
+                         json);
+  nr_free(json);
+  nr_php_package_destroy(package1);
+}
+
+static void test_php_packages_to_json_buffer(void) {
+  nrbuf_t* buf = nr_buffer_create(0, 0);
+  nr_php_packages_t* collection = nr_php_packages_create();
+  nr_php_package_t* package1;
+  nr_php_package_t* package2;
+  nr_php_package_t* package3;
+  nr_php_package_t* package4;
+  int count;
+
+  package1 = nr_php_package_create("Package One", "1.0.0");
+  // Add package with same key, but different value. Newer value will be kept
+  package2 = nr_php_package_create("Package One", "11.0");
+  package3 = nr_php_package_create("Package Two", "2.0.0");
+  // Add package with same key and same value. No action will happen
+  package4 = nr_php_package_create("Package Two", "2.0.0");
+
+  nr_php_packages_add_package(collection, package1);
+  nr_php_packages_add_package(collection, package2);
+  nr_php_packages_add_package(collection, package3);
+  nr_php_packages_add_package(collection, package4);
+
+  // Total package count should be 2 because two packages were duplicates with
+  // the same key
+  count = nr_php_packages_count(collection);
+  tlib_pass_if_int_equal("package count", 2, count);
+
+  // Test: adding packages to buffer
+  tlib_pass_if_bool_equal("filled collection bool check", true,
+                          nr_php_packages_to_json_buffer(collection, buf));
+
+  nr_buffer_add(buf, NR_PSTR("\0"));
+  tlib_pass_if_str_equal(
+      "filled collection",
+      "[[\"Package One\",\"11.0\",{}],[\"Package Two\",\"2.0.0\",{}]]",
+      nr_buffer_cptr(buf));
+  nr_php_packages_destroy(&collection);
+  nr_buffer_destroy(&buf);
+}
+
+static void test_php_packages_to_json(void) {
+  char* json;
+  nr_php_packages_t* h = nr_php_packages_create();
+  nr_php_package_t* package1;
+  nr_php_package_t* package2;
+
+  // Test: passing NULL does not crash
+  tlib_pass_if_null("NULL package", nr_php_packages_to_json(NULL));
+
+  // Test: convert all packages in hashmap to json
+  package1 = nr_php_package_create("Package One", "10.1.0");
+  package2 = nr_php_package_create("Package Two", "11.2.0");
```

Comment:
Add a package w/o a version to verify the JSON creation.

Reply:
added: [2d67c1b](https://github.com/newrelic/newrelic-php-agent/pull/754/commits/2d67c1bc407b743ecdbc9e6828bf42b03cdcecbc)